File: server/server-core/src/main/java/com/simple/castle/server/physic/unit/PhysicObject.java
package com.simple.castle.server.physic.unit;

import com.badlogic.gdx.physics.bullet.dynamics.btRigidBody;
import com.badlogic.gdx.physics.bullet.linearmath.btDefaultMotionState;
import com.badlogic.gdx.physics.bullet.linearmath.btMotionState;
import com.badlogic.gdx.utils.Disposable;

public abstract class PhysicObject implements Disposable {
    private final btRigidBody.btRigidBodyConstructionInfo constructionInfo;
    protected final btRigidBody body;
    protected final btMotionState motionState;

    PhysicObject(final btRigidBody.btRigidBodyConstructionInfo constructionInfo) {
        this.constructionInfo = constructionInfo;
        this.motionState = new btDefaultMotionState();
        this.body = new btRigidBody(constructionInfo);
        this.body.setMotionState(motionState);

        this.body.userData = this;
    }

    @Override
    public void dispose() {
        motionState.dispose();
        constructionInfo.dispose();
        body.dispose();
    }

    public btRigidBody getBody() {
        return body;
    }

    public btMotionState getMotionState() {
        return motionState;
    }
}

File: server/server-core/src/main/java/com/simple/castle/server/kt/physic/PhysicObject.kt
package com.simple.castle.server.kt.physic

import com.badlogic.gdx.physics.bullet.dynamics.btRigidBody
import com.badlogic.gdx.utils.Disposable
import com.simple.castle.server.kt.composition.MotionState

abstract class PhysicObject internal constructor(private val constructionInfo: btRigidBody.btRigidBodyConstructionInfo) : Disposable {
    val body: btRigidBody = btRigidBody(constructionInfo)
    val motionState: MotionState = MotionState()

    init {
        body.userData = this
    }

    override fun dispose() {
        motionState.dispose()
        constructionInfo.dispose()
        body.dispose()
    }
}